### PR TITLE
App deletion

### PR DESCRIPTION
**Pull request by archseer**

### CODE REVIEW

File: app.go [hunk 1]
```diff
@@ -133,6 +133,17 @@ func (r *AppRepo) Update(id string, data map[string]interface{}) (interface{}, e
...
 }
 
+func (r *AppRepo) Remove(id string) (interface{}, error) {
+	err := r.db.Exec("UPDATE apps SET deleted_at = now(), updated_at = current_timestamp WHERE app_id = $1 AND deleted_at IS NULL", id)
+	if err != nil {
+		return err
+	}
+
+	r.db.Exec("UPDATE releases SET deleted_at = now(), updated_at = current_timestamp WHERE app_id = $1 AND deleted_at IS NULL", id)
+	r.db.Exec("UPDATE app_resources SET deleted_at = now(), updated_at = current_timestamp WHERE app_id = $1 AND deleted_at IS NULL", id)
```
Comment (titanous) [inline app.go]: All of this should be in a transaction.

File: app.go [hunk 2]
```diff
@@ -133,6 +133,17 @@ func (r *AppRepo) Update(id string, data map[string]interface{}) (interface{}, e
 	return app, tx.Commit()
 }
 
+func (r *AppRepo) Remove(id string) (interface{}, error) {
+	err := r.db.Exec("UPDATE apps SET deleted_at = now(), updated_at = current_timestamp WHERE app_id = $1 AND deleted_at IS NULL", id)
+	if err != nil {
+		return err
+	}
+
+	r.db.Exec("UPDATE releases SET deleted_at = now(), updated_at = current_timestamp WHERE app_id = $1 AND deleted_at IS NULL", id)
```
Comment (titanous) [inline app.go]: Releases don't have an `app_id` column, so this is returning an error. Always make sure you check the error returned from calls, Go doesn't have exceptions and only panics for stuff like out of bounds access.

Comment (archseer) [inline app.go]: Oops, that was supposed to be `formations`.

File: app.go [hunk 3]
```diff
@@ -133,6 +133,17 @@ func (r *AppRepo) Update(id string, data map[string]interface{}) (interface{}, e
 	return app, tx.Commit()
 }
 
+func (r *AppRepo) Remove(id string) (interface{}, error) {
+	err := r.db.Exec("UPDATE apps SET deleted_at = now(), updated_at = current_timestamp WHERE app_id = $1 AND deleted_at IS NULL", id)
```
Comment (titanous) [inline app.go]: `now()` and `current_timestamp` are aliases for the same thing. Pick one.

Comment (titanous) [inline app.go]: Also, no need to set `updated_at` when deleting.

**Comment by jayd3e** [comment]
How hard would it be to add "listing apps" into this PR as well.

**Comment by titanous** [comment]
@jayd3e that already exists in the API and will be added to flynn-cli in flynn/flynn-cli#18.

**Comment by archseer** [comment]
No worries, I intend to rebase this later on.

**Comment by jayd3e** [comment]
@titanous :+1:

**Comment by titanous** [comment]
@archSeer is this ready for review?

**Comment by archseer** [comment]
@titanous Yes, it should be ready now. After you approve it, I'll flatten it via rebase, and do the CLI side.

File: controller_test.go [hunk 4]
```diff
@@ -220,6 +220,18 @@ func (s *S) TestUpdateApp(c *C) {
...
+func (s *S) TestDeleteApp(c *C) {
+	app := s.createTestApp(c, &ct.App{Name: "delete-app"})
+
+	path := "/apps/" + app.ID
+	res, err := s.Delete(path)
+	c.Assert(err, IsNil)
+	c.Assert(res.StatusCode, Equals, 200)
+
+	res, err = s.Get(path, app)
+	c.Assert(res.StatusCode, Equals, 404)
```
Comment (titanous) [inline controller_test.go:232]: We should test that an app with the same `name` can be subsequently created successfully.

File: app.go [hunk 5]
```diff
@@ -133,6 +133,31 @@ func (r *AppRepo) Update(id string, data map[string]interface{}) (interface{}, e
...
+		tx.Rollback()
+		return err
+	}
+
+	err = r.db.Exec("UPDATE formations SET deleted_at = now() WHERE app_id = $1 AND deleted_at IS NULL", id)
+	if err != nil {
+		tx.Rollback()
+		return err
+	}
+	err = r.db.Exec("UPDATE app_resources SET deleted_at = now() WHERE app_id = $1 AND deleted_at IS NULL", id)
```
Comment (titanous) [inline app.go]: These all need to be executed using the `tx`, not `r.db`.

Comment (archseer) [inline app.go]: Ooops.

File: app.go [hunk 6]
```diff
@@ -133,6 +133,31 @@ func (r *AppRepo) Update(id string, data map[string]interface{}) (interface{}, e
...
+		return err
+	}
+
+	err = tx.Exec("UPDATE apps SET deleted_at = now() WHERE app_id = $1 AND deleted_at IS NULL", id)
+	if err != nil {
+		tx.Rollback()
+		return err
+	}
+
+	err = tx.Exec("UPDATE formations SET deleted_at = now() WHERE app_id = $1 AND deleted_at IS NULL", id)
```
Comment (titanous) [inline app.go]: Also set `processes = NULL` so that the scheduler propagates the changes to the cluster.

**Comment by titanous** [comment]
LGTM. Address my last comment, squash, and then I'll merge.

**Comment by archseer** [comment]
There, fixed and rebased.

**Comment by titanous** [comment]
@archSeer It looks like the code is broken (see travis test results).

**Comment by titanous** [comment]
Still broken :/

**Comment by archseer** [comment]
Ah, one thing I forgot to do, is to change the create route to work wih app re-creation.

**Comment by archseer** [comment]
There we go, I'll do the same for artifacts in a separate PR.

**Comment by titanous** [comment]
Looks good, and the tests are passing. Please squash.

**Comment by archseer** [comment]
Squashed!

**Comment by titanous** [comment]
Nice work!

**Pull request merged**